## Introduction to Data Science

#### University of Redlands - DATA 101
#### Prof: Joanna Bieri [joanna_bieri@redlands.edu](mailto:joanna_bieri@redlands.edu)
#### [Class Website: data101.joannabieri.com](https://joannabieri.com/data101.html)

---------------------------------------
# Homework Day 8
---------------------------------------

GOALS:

1. Load data into Python that you find online
2. Understand data types and fix some errors
3. Find your own data to play with

----------------------------------------------------------

This homework has **5 questions** and **1 problem**.

NOTE: Be kind to yourself. Working with data can be hard! Every data set is different. **Seriously** come get help! Come to lab!


In [65]:
import numpy as np
import pandas as pd

import matplotlib.pyplot as plt
import plotly.express as px
from plotly.subplots import make_subplots
import plotly.io as pio
pio.renderers.defaule = 'colab'

You might need to install lxml. If you get a "ImportError: Missing optional dependency 'lxml'" Then uncomment and run the cell below

In [66]:
#!conda install -y lxml

## Try reading in some data - csv

Go to the [Cal Fire Website](https://www.fire.ca.gov/incidents) and scroll to the bottom to see the Incident Data. We will download the file named **ALL DATA AS CSV** this should put the data file into your Downloads folder. 

Next you need to move the file **mapdataall.csv** from your Downloads folder into your Day8 folder where you are doing your homework. You can open your Downloads folder and drag the file into JupyterLab side bar. Then I can run the command

    DF_raw = pd.read_csv('mapdataall.csv')

to load the data and look at the data frame.

In [67]:
DF = pd.read_csv("./mapdataall.csv")
DF

incident_name incident_is_final  incident_date_last_update  \
0       Bridge Fire                 Y       2018-01-09T13:46:00Z   
1         Pala Fire                 Y       2020-09-16T14:07:35Z   
2        River Fire                 Y       2022-10-24T11:39:23Z   
3     Fawnskin Fire                 Y       2013-04-22T09:00:00Z   
4         Gold Fire                 Y       2013-05-01T07:00:00Z   
...             ...               ...                        ...   
3335      Lost Fire                 Y       2025-09-28T09:51:03Z   
3336   Martell Fire                 Y       2025-09-28T09:49:59Z   
3337      Cobb Fire                 Y       2025-09-29T07:12:14Z   
3338     Vleck Fire                 N  2025-09-30T13:00:14-07:00   
3339    Paloma Fire                 Y       2025-09-30T21:41:00Z   

     incident_date_created          incident_administrative_unit  \
0     2017-10-31T11:22:00Z       Shasta-Trinity National Forest    
1     2009-05-24T14:56:00Z               CAL FIRE San Diego Unit   
2     2013-02-24T08:16:00Z          CAL FIRE San Bernardino Unit   
3     2013-04-20T17:30:00Z        San Bernardino National Forest   
4     2013-04-30T12:59:00Z  CAL FIRE Madera-Mariposa-Merced Unit   
...                    ...                                   ...   
3335  2025-09-25T08:00:00Z        Ventura County Fire Department   
3336  2025-09-26T14:45:08Z      CAL FIRE Tuolumne-Calaveras Unit   
3337  2025-09-26T15:56:10Z            Six Rivers National Forest   
3338  2025-09-29T12:02:36Z        CAL FIRE Amador-El Dorado Unit   
3339  2025-09-30T01:08:26Z  CAL FIRE Madera-Mariposa-Merced Unit   

      incident_administrative_unit_url incident_county  \
0                                  NaN          Shasta   
1                                  NaN       San Diego   
2                                  NaN            Inyo   
3                                  NaN  San Bernardino   
4                                  NaN          Madera   
...                                ...             ...   
3335                               NaN         Ventura   
3336                               NaN      Stanislaus   
3337                               NaN        Humboldt   
3338                               NaN      Sacramento   
3339                               NaN        Mariposa   

                                      incident_location  \
0     I-5 and Turntable Bay, 7 miles NE of Shasta Lake    
1           Hwy 76 and Pala Temecula, northwest of Pala   
2     south of Narrow Gauge Rd & north of Hwy 136, e...   
3     west of Delamar Mountain, north of the communi...   
4     Between Road 210 and Road 200 near Fine Gold C...   
...                                                 ...   
3335                Near Lost Canyon Drive, Simi Valley   
3336  South of Telegraph Road and Northeast of Sonor...   
3337    Cobb Road and Rattlesnake Bridge Road, Dinsmore   
3338                 Van Vleck Road and Jackson Highway   
3339                      Paloma Road and Hornitos Road   

      incident_acres_burned  incident_containment  ... incident_latitude  \
0                      37.0                 100.0  ...         40.774000   
1                     122.0                 100.0  ...          1.000000   
2                     407.0                 100.0  ...         36.602575   
3                      30.0                 100.0  ...         34.288877   
4                     274.0                 100.0  ...         37.116295   
...                     ...                   ...  ...               ...   
3335                   13.0                 100.0  ...         34.301578   
3336                   37.0                 100.0  ...         37.906127   
3337                   12.0                 100.0  ...         40.463864   
3338                   30.0                  95.0  ...         38.482401   
3339                   86.0                 100.0  ...         37.482008   

     incident_type                           incident

**Q1** How many variables and observations?

**Q2** How many different incident types are there?

**Q3** Make a histogram of the acres burned and color the bars by whether or not the incident was a calfire incident. You will probably need to make a mask to remove very small and very large fires. How many fires burned more than 100,000 acres? What is the largest fire in the data?

**Extra Q** EXTRA - CHALLENGE - See if you can create a graph that answers the question: Are fires getting bigger or more frequent over time? You get complete creative control on how to answer this question!

In [68]:
# You will need to write some code to answer the questions.
obv = DF.shape[0]
var = DF.shape[1]

print(f"The number of observations is {obv}")
print(f"The number of variables is {var}")

The number of observations is 3340
The number of variables is 23


In [69]:
#DF.columns
diff_incidents_type = DF['incident_type'].unique()
number_dit = len(diff_incidents_type)
print(f"Number of different incident types: {number_dit}")

Number of different incident types: 5


In [70]:
mask_acres = (DF['incident_acres_burned'] > 100000)
DF_acres = DF[mask_acres]

acre_obv = DF_acres.shape[0]
print(f"There are {acre_obv} fires that burned more than 100,000 acres")

There are 21 fires that burned more than 100,000 acres


In [71]:
filtered_df = DF[(DF['incident_acres_burned'] >= 100) & (DF['incident_acres_burned'] <= 10000)]

fig = px.histogram(filtered_df,
                   x='incident_name',
                   y='incident_acres_burned',
                   color='calfire_incident')

fig.update_layout(xaxis_tickangle=-45)
fig.show()

In [72]:
max_burned = DF_acres['incident_acres_burned'].max()
mask_burned = (DF_acres['incident_acres_burned'] == max_burned)

DF_mask_burned = DF_acres[mask_burned]
print(DF_mask_burned['incident_name'])

1653    August Complex (includes Doe Fire)
Name: incident_name, dtype: object


## Try reading in some data from Wikipedia - html

Here we will explore academy award winning films. Go to the [Wiki for the List of Academy Award Winning Films](https://en.wikipedia.org/wiki/List_of_Academy_Award%E2%80%93winning_films). Look at what type of data is there. How many tables? Any weird looking data?

Now we will read the html data into Python and show the data in DF[0]

In [73]:
# Here is some helper code
# This is code that will read in the data and then fix the Year column

#####
'''
NOTE: Many websites try to stop you from scraping even simple tables. 
    In just one year wikipedia made big changes... so I changed the code here.
    We can talk more about this code in class.
'''
import requests
headers = {
    'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/91.0.4472.124 Safari/537.36'
}
my_website = "https://en.wikipedia.org/wiki/List_of_Academy_Award%E2%80%93winning_films"
response = requests.get(my_website, headers=headers)
#####

# Normal pandas code
DF = pd.read_html(response.content)
DF_raw = DF[0]
DF_raw['Year'] = DF_raw['Year'].apply(lambda x: int(x.split('/')[0]))
DF_raw

Film  Year Awards  \
0                                                 Anora  2024      5   
1                                         The Brutalist  2024      3   
2                                          Emilia Pérez  2024      2   
3                                                Wicked  2024      2   
4                                        Dune: Part Two  2024      2   
...                                                 ...   ...    ...   
1382                            The Yankee Doodle Mouse  1943      1   
1383                                       The Yearling  1946      2   
1384  Yesterday, Today and Tomorrow (Ieri, oggi, dom...  1964      1   
1385                         You Can't Take It with You  1938      2   
1386                    Zorba the Greek (Alexis Zorbas)  1964      3   

     Nominations  
0              6  
1             10  
2             13  
3             10  
4              5  
...          ...  
1382           1  
1383           7  
1384           1  
1385           7  
1386           7  

[1387 rows x 4 columns]

**Q4** Following along with the lecture notes or video, fix the data in the 'Awards' column.

In [74]:
DF_raw['Awards'] = DF_raw['Awards'].apply(lambda x: int(x.split('(')[0]))

awards = DF_raw['Awards'].value_counts()
DF_award = awards.reset_index().rename(columns={"index":"value", 0: "count"})
DF_award

Awards  count
0        1    996
1        2    160
2        3     79
3        4     48
4        0     36
5        5     27
6        7     13
7        6     13
8        8      8
9       11      3
10       9      3
11      10      1

**Q5** Now try to fix the data in the "Nominations" column - see if you can do it without looking at the answer.

In [75]:
DF_raw['Nominations'] = DF_raw['Nominations'].apply(lambda x: int(x.split('[')[0]))

noms = DF_raw['Nominations'].value_counts()
DF_noms = noms.reset_index().rename(columns={"index":"value", 0: "count"})
DF_noms

Nominations  count
0             1    557
1             2    143
2             4    115
3             5    106
4             3    105
5             6     81
6             7     75
7             8     62
8            10     42
9             9     31
10           11     24
11           12     17
12            0     14
13           13     12
14           14      3

In [76]:
DF_raw.dtypes

Film           object
Year            int64
Awards          int64
Nominations     int64
dtype: object

## Problem 1

Your homework today will be to see if you can find some data of your own. This can be the first steps you take toward your final project. 

You should:

* Find some data online
* Read that data into Python using the Pandas commands we learned
* Look at the DataFrame - number of variables, number of observations, AND the dtypes. Comment on what you see.
* Try to do summary statistics (.describe()). Does it work like expected?
* Attempt to fix any data, or explain why the data does not need to be fixed.
* Make some sort of graph using columns in your data.

In [77]:
NBA_DF = pd.read_csv("./nba2023.csv")
NBA_DF

PName  POS Team  Age  GP   W   L     Min   PTS  FGM  \
0               Jayson Tatum   SF  BOS   25  74  52  22  2732.2  2225  727   
1                Joel Embiid    C  PHI   29  66  43  23  2284.1  2183  728   
2                Luka Doncic   PG  DAL   24  66  33  33  2390.5  2138  719   
3    Shai Gilgeous-Alexander   PG  OKC   24  68  33  35  2416.0  2135  704   
4      Giannis Antetokounmpo   PF  MIL   28  63  47  16  2023.6  1959  707   
..                       ...  ...  ...  ...  ..  ..  ..     ...   ...  ...   
534         Alondes Williams  NaN  BKN   23   1   1   0     5.3     0    0   
535            Deonte Burton  NaN  SAC   29   2   1   1     6.5     0    0   
536            Frank Jackson  NaN  UTA   24   1   0   1     5.0     0    0   
537       Michael Foster Jr.  NaN  PHI   20   1   1   0     1.0     0    0   
538           Sterling Brown  NaN  LAL   28   4   2   2    24.4     0    0   

     ...  REB  AST  TOV  STL  BLK   PF    FP  DD2  TD3  +/-  
0    ...  649  342  213   78   51  160  3691   31    1  470  
1    ...  670  274  226   66  112  205  3706   39    1  424  
2    ...  569  529  236   90   33  166  3747   36   10  128  
3    ...  329  371  192  112   65  192  3425    3    0  149  
4    ...  742  359  246   52   51  197  3451   46    6  341  
..   ...  ...  ...  ...  ...  ...  ...   ...  ...  ...  ...  
534  ...    1    0    2    0    0    1    -1    0    0   -5  
535  ...    0    0    0    0    0    0     0    0    0    5  
536  ...    2    1    0    0    0    0     4    0    0   -2  
537  ...    0    0    0    0    0    0     0    0    0   -2  
538  ...    8    2    0    3    0    4    22    0    0   -4  

[539 rows x 30 columns]

In [78]:
NBA_obv = NBA_DF.shape[0]
NBA_var = NBA_DF.shape[1]

print(f"The number of observations is {NBA_obv}")
print(f"The number of variables is {NBA_var}")

The number of observations is 539
The number of variables is 30


In [79]:
NBA_DF.dtypes

PName     object
POS       object
Team      object
Age        int64
GP         int64
W          int64
L          int64
Min      float64
PTS        int64
FGM        int64
FGA        int64
FG%      float64
3PM        int64
3PA        int64
3P%      float64
FTM        int64
FTA        int64
FT%      float64
OREB       int64
DREB       int64
REB        int64
AST        int64
TOV        int64
STL        int64
BLK        int64
PF         int64
FP         int64
DD2        int64
TD3        int64
+/-        int64
dtype: object

In [80]:
NBA_DF.describe()

Age          GP           W           L          Min  \
count  539.000000  539.000000  539.000000  539.000000   539.000000   
mean    25.970315   48.040816   24.018553   24.022263  1103.617625   
std      4.315513   24.650686   14.496366   13.445866   827.765114   
min     19.000000    1.000000    0.000000    0.000000     1.000000   
25%     23.000000   30.500000   12.000000   14.000000   329.000000   
50%     25.000000   54.000000   25.000000   25.000000   970.200000   
75%     29.000000   68.000000   36.000000   34.000000  1845.900000   
max     42.000000   83.000000   57.000000   60.000000  2963.200000   

               PTS         FGM          FGA         FG%         3PM  ...  \
count   539.000000  539.000000   539.000000  539.000000  539.000000  ...   
mean    523.426716  191.576994   403.005566   46.325232   56.324675  ...   
std     498.084360  178.351286   369.595909   10.967271   60.916821  ...   
min       0.000000    0.000000     0.000000    0.000000    0.000000  ...   
25%     120.500000   45.500000    93.500000   41.650000    5.000000  ...   
50%     374.000000  138.000000   300.000000   45.500000   36.000000  ...   
75%     769.500000  283.500000   598.500000   50.600000   92.000000  ...   
max    2225.000000  728.000000  1559.000000  100.000000  301.000000  ...   

              REB         AST         TOV         STL         BLK          PF  \
count  539.000000  539.000000  539.000000  539.000000  539.000000  539.000000   
mean   198.254174  115.545455   61.300557   33.270872   21.241187   91.181818   
std    181.819962  129.578453   58.279185   28.336745   26.529238   66.206731   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%     50.500000   22.000000   14.500000    8.500000    5.000000   32.000000   
50%    159.000000   69.000000   44.000000   28.000000   13.000000   86.000000   
75%    286.000000  162.500000   92.500000   51.000000   28.000000  140.000000   
max    973.000000  741.000000  300.000000  128.000000  193.000000  279.000000   

                FP         DD2         TD3         +/-  
count   539.000000  539.000000  539.000000  539.000000  
mean   1036.938776    4.011132    0.220779    0.000000  
std     894.081896    8.770932    1.564432  148.223909  
min      -1.000000    0.000000    0.000000 -642.000000  
25%     254.000000    0.000000    0.000000  -70.000000  
50%     810.000000    0.000000    0.000000   -7.000000  
75%    1646.000000    3.000000    0.000000   57.000000  
max    3842.000000   65.000000   29.000000  640.000000  

[8 rows x 27 columns]

In [81]:
young = NBA_DF.copy()
mask = (NBA_DF['Age'] <= 25) & (NBA_DF['PTS'] >= 1230)
young = young[mask]

young['PPG'] = young['PTS'] / 82
young['APG'] = young['AST'] / 82
young['RPG'] = young['REB'] / 82
young['SPG'] = young['STL'] / 82
young['BPG'] = young['BLK'] / 82

columns = ['PName', 'Age', 'PPG', 'APG', 'RPG', 'SPG', 'BPG']
young = young[columns]

young.rename(columns={'PName':'Player'}, inplace=True)

young.sort_values(by="Age", ascending=False)

Player  Age        PPG       APG       RPG       SPG  \
0              Jayson Tatum   25  27.134146  4.170732  7.914634  0.951220   
11             De'Aaron Fox   25  22.268293  5.451220  3.731707  1.012195   
26              Bam Adebayo   25  18.646341  2.926829  8.390244  1.073171   
15          Lauri Markkanen   25  20.621951  1.500000  6.951220  0.512195   
2               Luka Doncic   24  26.073171  6.451220  6.939024  1.097561   
58             Desmond Bane   24  15.207317  3.097561  3.548780  0.682927   
8                Trae Young   24  23.341463  9.036585  2.646341  0.975610   
3   Shai Gilgeous-Alexander   24  26.036585  4.524390  4.012195  1.365854   
18             Jordan Poole   23  20.426829  4.500000  2.743902  0.768293   
23                Ja Morant   23  19.463415  6.012195  4.353659  0.804878   
39           Keldon Johnson   23  16.890244  2.231707  3.878049  0.560976   
43              Tyler Herro   23  16.426829  3.414634  4.390244  0.621951   
48          Anfernee Simons   23  15.926829  3.085366  1.963415  0.512195   
31           Darius Garland   23  18.170732  6.560976  2.256098  1.036585   
38               RJ Barrett   22  17.451220  2.451220  4.487805  0.378049   
17              Jalen Green   21  20.524390  3.426829  3.463415  0.719512   
5           Anthony Edwards   21  23.731707  4.268293  5.585366  1.524390   
32             Franz Wagner   21  18.109756  3.451220  4.012195  0.939024   
53              Evan Mobley   21  15.573171  2.731707  8.670732  0.731707   
37           Paolo Banchero   20  17.524390  3.280488  6.060976  0.731707   
49       Bennedict Mathurin   20  15.878049  1.414634  3.878049  0.585366   
57              Josh Giddey   20  15.365854  5.719512  7.304878  0.695122   

         BPG  
0   0.621951  
11  0.280488  
26  0.743902  
15  0.463415  
2   0.402439  
58  0.268293  
8   0.109756  
3   0.792683  
18  0.256098  
23  0.195122  
39  0.134146  
43  0.195122  
48  0.170732  
31  0.109756  
38  0.182927  
17  0.219512  
5   0.707317  
32  0.207317  
53  1.451220  
37  0.475610  
49  0.158537  
57  0.378049

In [82]:
px.histogram(young,
              x = 'Player',
              y = 'PPG',
             labels = {'PPG': 'PPG NBA 2023'},
              color = 'Age'
            )

**I reduced the number of observations from 539 to 22, filtering out players based on age (looking at players 25 or younger) and then looking at which players averaged more than 15 points per game in the 2023 NBA season. ALl the data seems to be behaving correctly as there are no notable outliers in the new DataFrame focusing on young rising superstars. Everything works as expected, especially since this is a smaller dataset once we filtered out what we wanted, so there is less room for error.**